### PR TITLE
New analyzer pass to simplify case when

### PR DESCRIPTION
**Pull request by liuneng1994**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Simply case when clause to get better performance

In some case we can simplify case when to use primary key index.
example:
```
select * from test where (case x when 'a' then 1 when 'b' then 2 else 3 end) = 1
```
=>
```
select * from test where x = 'a'
```

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit 9ad070e417e0ff0e2f692555ff2727a8c9997bbc with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/61919/9ad070e417e0ff0e2f692555ff2727a8c9997bbc/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Comment by liuneng1994** [comment]
There are too many bugs and I need to reconsider the implementation method.

File: src/Analyzer/Passes/CaseWhenSimplifyPass.cpp [hunk 1]
```diff
@@ -0,0 +1,479 @@
...
+        if (has_else)
+        {
+            else_node = case_args.back()->as<ConstantNode>();
+            if (!else_node)
+            {
+                failed = true;
+                return;
+            }
+            if (else_node->getValue().isNull())
+                has_else = false;
```
Comment (novikd) [inline src/Analyzer/Passes/CaseWhenSimplifyPass.cpp:118]: Why?

**Review by novikd — CHANGES_REQUESTED** [review]
The provided test doesn't use the new infrastructure. Please, fix it. Also, I think it's useful to add a performance test.
The implementation requires to have more comments because it's hard to read now.

**Comment by liuneng1994** [comment]
Sorry, this optimization still requires a lot of work to be perfect. I don’t have more energy to spend on it at the moment.

**Pull request closed without merging**